[server] Verify TLS certificate in LDAPS connections

jimis opened the pull request:
If LDAP authentication is used in the server and the LDAP server
connection is using LDAPS, then reject the connection if the certificate
is not trusted by the system.

This can be overriden by setting the environment variable
LDAPTLS_REQCERT=never just like in the ldapsearch command.

NOTE: The option used before to skip the SSL certificate verification
was OPT_X_TLS_ALLOW, which is documented in python-ldap as "used
internally by slapd server". Thus the one used now is OPT_X_TLS_NEVER.

Resolves #3082

csordasmarton commented on web/server/codechecker_server/auth/cc_ldap.py:
```diff
@@ -238,7 +244,8 @@ def __init__(self, ldap_config, who=None, cred=None):
 
         # Check cert if available but do not fail if not.
-        ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_ALLOW)
+        if os.environ.get("LDAPTLS_REQCERT") == "never":
```
I am thinking to move this option to the config file instead of environment variable similar to `deref` config option and others:
https://github.com/Ericsson/codechecker/blob/master/web/server/config/server_config.json#L31

The name can be for example `tls_require_cert`.
The handled values can be `never` and `allow`.
The default value can be `allow` to not change the behavior of the server. So who already use it, they will not need to change anything.

Also extend the documentation with the new config option: https://github.com/Ericsson/codechecker/blob/master/docs/web/authentication.md

jimis replied:
@csordasmarton according to the [python-ldap documentation](https://www.python-ldap.org/en/python-ldap-3.3.0/reference/ldap.html#ldap.OPT_X_TLS_REQUIRE_CERT) the value `allow` that I removed is only to be used internally by the slapd server. So I don't think I should re-introduce it. From our perspective, it seems to be equivalent to "never" since it allows insecure connections to go through.

I suggest the new option only handles one value: `never`. In that case no certificate verification happens, like today. In every other case, no option is passed to the ldap library using OPT_X_TLS_REQUIRE_CERT (which means secure behaviour).

This will break existing insecure LDAPS setups (will require admins to modify configuration to add `tls_require_cert="never"`) but it's probably better to do it sooner rather than later, since the ldap module doesn't seem widely used yet (it's missing core functionality for us at least).

jimis commented:
Rebased and pushed update according to my comment above (default behaviour is secure, compatibility is broken). I also fixed an irrelevant typo that kind of proves my point: the LDAP module is not widely used yet, so it's better to break compatibility at this early stage and enforce secure LDAPS connections by default.

csordasmarton commented on docs/web/authentication.md:
```diff
@@ -182,6 +182,11 @@ servers as it can elongate the authentication process.
 
+ * `tls_require_cert`
+
+   If set to "never", skip verification of certificate in LDAPS connections
+   (insecure!).
```
Just add some highlight:
```suggestion
   If set to `never`, skip verification of certificate in LDAPS connections
   (!!! INSECURE !!!).
```

csordasmarton commented on docs/web/authentication.md:
```diff
@@ -268,7 +273,8 @@ servers as it can elongate the authentication process.
     {
-      "connection_url" : "ldaps://secure.internal.example.org:636",
+      "connection_url"   : "ldaps://secure.internal.example.org:636",
+      "tls_require_cert" : "never",
```
You are right. The default behavior shouldn't be insecure. So I think we need to change this value to `null˙ here yes?

csordasmarton reviewed (changes requested):
I have just some tiny comments otherwise LGTM!

jimis commented:
Updated, thanks for the review.

csordasmarton reviewed (approved):
LGTM! Thank you for this patch :blush:
